### DISCUSSION TITLE
Different decision for mime_content_type() and \Safe\mime_content_type()

### DISCUSSION THREAD

**Discussion by OskarStark** [composer-unused/composer-unused#101] [Ideas]
## Is your feature request related to a problem? Please describe
Yes and I am not really sure if this is a bug or a not yet implemented feature?

Lets say I have:
```php
<?php

namespace App\Collector;

final class FileCollector
{
    // ...

    public function getMimeType(): string
    {
       return mime_content_type($tempfilepathname);
    }
}
```
and the following `composer.json`:
```json
    "require": {
        "php": "^7.4",
        "ext-fileinfo": "*",
       // ...
    },
```

everything is fine.
Now I require `thecodingmachine/safe` and change my code accordingly:
```dif
    public function getMimeType(): string
    {
-       return mime_content_type($tempfilepathname);
+       return \Safe\mime_content_type($tempfilepathname);
    }
}
```

**now composer unused reports `ext-fileinfo` as unused** 😢 

This is the same behavior for alle \Safe\-wrapped methods.

## Describe the solution you'd like
I don't have a clear idea how to fix it unless having special scanners for some libs, like PHPStan does.

## Describe alternatives you've considered
 n/a

## Teachability, Documentation, Adoption, Migration Strategy
 n/a

cc @localheinz

**Comment by icanhazstring** [comment]
Can you try this against #89 using `composer-unused -x`?

**Comment by icanhazstring** [comment]
Hm. Looking at the `composer.json` of `thecodingmachine/safe` it seems that `ext-fileinfo` is not event suggested.
This means that `composer-unused` will most likely report this extension as unused.

As I only scan for root symbols (from the project/lib itself) or foreign symbols from required packages.
But currently, I do not scan used symbols from each of the required packages to match them against required packages of the root.

One fix would be that `thecodingmachine/safe` should have the extension as `suggest`. Or you could add the extension to be ignored by `composer-unused` inside your own `composer.json`

**Comment by OskarStark** [comment]
> Hm. Looking at the composer.json of thecodingmachine/safe it seems that ext-fileinfo is not event suggested.

&

>One fix would be that thecodingmachine/safe should have the extension as suggest. 

Thats true, otherwise they have to suggest **all** extensions, which would the result in "No ext-* will be ever marked as unused, if you require the lib", right?

> But currently, I do not scan used symbols from each of the required packages to match them against required packages of the root.

That make sense, but I would love to have custom scanners when using special libs.

> Or you could add the extension to be ignored by composer-unused inside your own composer.json

I can do this, but its wrong from my POV, one would not understand why I `require` **and** ignore it by `unused` at the same time 🤔 

Thanks for your time 😃  and your tool ❤️

**Comment by icanhazstring** [comment]
> I can do this, but its wrong from my POV, one would not understand why I require and ignore it by unused at the same time 🤔

True. That would seem weird.

> That make sense, but I would love to have custom scanners when using special libs.

Hopefully, I can improve the code further in the next week when I got some time on my hand.
Also adding the ability to add customer scanners is one thing that is on my list for a long time.
That would solve a lot of problems e.g. non PHP configuration files.

I still didn't find the time on how to get something like this done :)

**Comment by icanhazstring** [comment]
Related to #69
